src/ExtendedXmlSerializer/ExtendedXmlSerializer.cs

```diff
@@ -378,44 +386,35 @@ private object ReadXmlDictionary(XElement currentNode, TypeDefinition type, obje
 
         private object ReadXmlArray(XElement currentNode, TypeDefinition type, object instance = null)
         {
-            int arrayCount = currentNode.Elements().Count();
             var elements = currentNode.Elements().ToArray();
-
+            int arrayCount = elements.Length;
             object list = null;
             Array array = null;
             if (type.IsArray)
             {
-                array = (Array) instance ?? Array.CreateInstance(type.Type.GetElementType(), arrayCount);
+                array =  instance as Array ?? Array.CreateInstance(type.Type.GetElementType(), arrayCount);
             }
             else
             {
                 list = instance ?? type.ObjectActivator();
             }
+
+            var elementType = ElementTypeLocator.Default.Locate( type.Type );
+            var elementDefinition = TypeDefinitionCache.GetDefinition(elementType);
             for (int i = 0; i < arrayCount; i++)
             {
                 var element = elements[i];
-                TypeDefinition cd = null;
-                var ta = element.Attribute("type");
-                if (ta != null)
-                {
-                    var currentNodeType = TypeDefinitionCache.GetType(ta.Value);
-                    cd = TypeDefinitionCache.GetDefinition(currentNodeType);
-                }
+                var ta = element.Attribute(Type);
+                var definition = ta != null ? TypeDefinitionCache.GetDefinition( TypeDefinitionCache.GetType(ta.Value) ) : elementDefinition;
```

Comment:
I went ahead and submitted a new commit for this, since I was in the code for the other commits.